packages/mui-base/src/useTransition/useTransition.ts

```diff
@@ -0,0 +1,104 @@
+import * as React from 'react';
+import { TransitionContext, TransitionContextValue } from './TransitionContext';
+/**
+ *
+ * API:
+ *
+ * - [useTransition API](https://mui.com/base-ui/api/use-transition/)
+ */
+export function useTransition() {
+  const transitionContext = React.useContext(TransitionContext);
+  if (!transitionContext) {
+    throw new Error('Missing transition context');
+  }
+
+  const {
+    registerTransition,
+    requestEnter: requestOpen,
+    onEntering,
+    onEntered,
+    onExiting,
+    onExited,
+    hasExited,
+  } = transitionContext;
+
+  React.useEffect(() => {
+    return registerTransition();
+  }, [registerTransition]);
+
+  return {
+    onEntering,
+    onEntered,
+    onExiting,
+    onExited,
+    requestedEnter: requestOpen,
+    hasExited,
+  };
+}
+
+export function useTransitionableElement(requestEnter: boolean) {
```

Comment:
I tried it in https://github.com/mui/material-ui/pull/39326/commits/9339c3ff9705d24f8f46610fb8cf5bd069a40c31, but I'm not sure it's better. There's more to reason about and the code is much longer.
What do you think?

Reply:
I think it's better with a reducer as the `hasEntered` and `inTransition` are heavily related. With the reducer, it will be harder for them to become unsynced. It's a personal preference, though, and if there are not many changes to the hook in the future, my fear of them becoming unsynced by mistake is low and coverable by tests. I'll leave the decision to you 🤗